File: apps/web-v2/src/lib/feed/mosaic-layout.test.ts
import { describe, expect, test } from 'bun:test';

import {
  MOSAIC_FALLBACK_LAYOUT,
  MOSAIC_LAYOUT_MODES,
  captionWordLimit,
  isPagedLayout,
  mosaicAspectRatio,
  mosaicTiles,
  resolveMosaicLayout,
  tileCarriesCaption,
  type MosaicTile,
} from './mosaic-layout';

describe('resolveMosaicLayout — l’agencement CHOISI par l’auteur, lu sur `storyEffects` (#6514)', () => {
  test('une valeur connue sur un document canvas v3 est servie telle quelle', () => {
    expect(resolveMosaicLayout({ v: 3, scenes: [], layout: 'hero' })).toBe('hero');
    expect(resolveMosaicLayout({ v: 3, layout: 'wave' })).toBe('wave');
    expect(resolveMosaicLayout({ v: 3, layout: 'reel' })).toBe('reel');
    expect(resolveMosaicLayout({ v: 3, layout: 'sine' })).toBe('sine');
    expect(resolveMosaicLayout({ v: 3, layout: 'carousel' })).toBe('carousel');
  });

  test('`layout` ABSENT ⇒ le carrousel — toute publication antérieure au 2026-09-06', () => {
    expect(resolveMosaicLayout({ v: 3, scenes: [] })).toBe('carousel');
    expect(MOSAIC_FALLBACK_LAYOUT).toBe('carousel');
  });

  test('`layout` INCONNU (écrit par un client plus récent) ⇒ le carrousel, jamais une panne', () => {
    expect(resolveMosaicLayout({ v: 3, layout: 'spiral' })).toBe('carousel');
    expect(resolveMosaicLayout({ v: 3, layout: 7 })).toBe('carousel');
    expect(resolveMosaicLayout({ v: 3, layout: '' })).toBe('carousel');
  });

  test('aucun `storyEffects` servi ⇒ le carrousel', () => {
    expect(resolveMosaicLayout(undefined)).toBe('carousel');
    expect(resolveMosaicLayout(null)).toBe('carousel');
    expect(resolveMosaicLayout('hero')).toBe('carousel');
  });

  /** Miroir de `StoryEffects` (iOS) : seul un document MARQUÉ `v >= 3` est un
   * canvas, et seul un canvas porte `layout` — un blob v1 n'en a jamais eu. */
  test('la MARQUE décide : un blob sans `v >= 3` ne porte aucun agencement, un rang supérieur se lit', () => {
    expect(resolveMosaicLayout({ layout: 'hero' })).toBe('carousel');
    expect(resolveMosaicLayout({ v: 2, layout: 'hero' })).toBe('carousel');
    expect(resolveMosaicLayout({ v: 4, layout: 'sine' })).toBe('sine');
  });

  test('les cinq valeurs sont exactement celles de `MosaicLayoutMode` — le carrousel seul pagine', () => {
    expect([...MOSAIC_LAYOUT_MODES].sort()).toEqual(['carousel', 'hero', 'reel', 'sine', 'wave']);
    expect(MOSAIC_LAYOUT_MODES.filter(isPagedLayout)).toEqual(['carousel']);
  });
});

const rounded = (tiles: readonly MosaicTile[]) =>
  tiles.map((t) => ({
    index: t.index,
    x: Math.round(t.x * 1000) / 1000,
    y: Math.round(t.y * 1000) / 1000,
    width: Math.round(t.width * 1000) / 1000,
    height: Math.round(t.height * 1000) / 1000,
    overflow: t.overflow,
  }));

describe('mosaicTiles — la géométrie de `MosaicLayout.swift`, en fractions de la boîte', () => {
  test('aucun visuel ⇒ aucune tuile ; un seul ⇒ une tuile pleine, quel que soit le mode', () => {
    expect(mosaicTiles(0, 'hero')).toEqual([]);
    for (const mode of MOSAIC_LAYOUT_MODES) {
      expect(rounded(mosaicTiles(1, mode))).toEqual([{ index: 0, x: 0, y: 0, width: 1, height: 1, overflow: 0 }]);
    }
  });

  test('`hero` à trois : une grande tuile à 0,62, deux satellites en colonne', () => {
    expect(rounded(mosaicTiles(3, 'hero'))).toEqual([
      { index: 0, x: 0, y: 0, width: 0.62, height: 1, overflow: 0 },
      { index: 1, x: 0.634, y: 0, width: 0.366, height: 0.493, overflow: 0 },
      { index: 2, x: 0.634, y: 0.507, width: 0.366, height: 0.493, overflow: 0 },
    ]);
  });

  test('`wave` à quatre : même largeur, hauteurs qui ondulent, centrées', () => {
    expect(rounded(mosaicTiles(4, 'wave'))).toEqual([
      { index: 0, x: 0, y: 0, width: 0.24, height: 1, overflow: 0 },
      { index: 1, x: 0.254, y: 0.13, width: 0.24, height: 0.74, overflow: 0 },
      { index: 2, x: 0.507, y: 0, width: 0.24, height: 1, overflow: 0 },
      { index: 3, x: 0.761, y: 0.13, width: 0.24, height: 0.74, overflow: 0 },
    ]);
  });

  test('`sine` à trois : une en haut, une en bas, une en haut', () => {
    expect(rounded(mosaicTiles(3, 'sine')).map((t) => t.y)).toEqual([0, 0.38, 0]);
    expect(rounded(mosaicTiles(3, 'sine')).every((t) => t.height === 0.62)).toBe(true);
  });

  test('`reel` DÉBORDE à droite : des tuiles de 0,60 au pas de 0,628', () => {
    expect(rounded(mosaicTiles(3, 'reel')).map((t) => [t.x, t.width, t.height])).toEqual([
      [0, 0.6, 1],
      [0.628, 0.6, 1],
      [1.256, 0.6, 1],
    ]);
  });

  test('au-delà de quatre, les mosaïques plafonnent et comptent le RESTE sur la dernière tuile seule', () => {
    const tiles = mosaicTiles(6, 'hero');
    expect(tiles).toHaveLength(4);
    expect(tiles.map((t) => t.overflow)).toEqual([0, 0, 0, 2]);
  });

  test('le carrousel ne plafonne pas et ne cache rien', () => {
    const tiles = mosaicTiles(6, 'carousel');
    expect(tiles).toHaveLength(6);
    expect(tiles.every((t) => t.overflow === 0 && t.width === 1 && t.height === 1)).toBe(true);
  });

  test('chaque mosaïque déclare le rapport hauteur / largeur de sa boîte', () => {
    expect(mosaicAspectRatio('wave')).toBe(0.78);
    expect(mosaicAspectRatio('hero')).toBe(0.82);
    expect(mosaicAspectRatio('reel')).toBe(1.05);
    expect(mosaicAspectRatio('sine')).toBe(0.92);
  });
});

describe('la légende d’une tuile — là où la place le permet, en MOTS', () => {
  test('seules la grande tuile d’un `hero`, les tuiles d’un `reel` et un visuel seul portent une légende', () => {
    expect(tileCarriesCaption('hero', 0, 3)).toBe(true);
    expect(tileCarriesCaption('hero', 1, 3)).toBe(false);
    expect(tileCarriesCaption('reel', 2, 3)).toBe(true);
    expect(tileCarriesCaption('wave', 0, 3)).toBe(false);
    expect(tileCarriesCaption('sine', 0, 3)).toBe(false);
    expect(tileCarriesCaption('wave', 0, 1)).toBe(true);
  });

  test('le budget de mots se réduit à la part de la rangée, jamais sous deux mots', () => {
    expect(captionWordLimit('hero', 3, 0.62)).toBe(5);
    expect(captionWordLimit('reel', 3, 0.6)).toBe(12);
    expect(captionWordLimit('wave', 4, 0.2395)).toBe(2);
    expect(captionWordLimit('hero', 1, 1)).toBe(20);
  });
});

File: apps/android/sdk-ui/src/test/kotlin/me/meeshy/ui/component/media/MosaicLayoutTest.kt
package me.meeshy.ui.component.media

import com.google.common.truth.Truth.assertThat
import me.meeshy.sdk.model.MosaicLayoutMode
import org.junit.Test

/**
 * La géométrie des agencements d'une publication (#6514), cotes reprises de
 * `MosaicLayout.swift` (`packages/MeeshySDK/Sources/MeeshyUI/Story/`) — la
 * même que celle que `apps/web/src/lib/feed/mosaic-layout.ts` dérive. Les
 * cadres sont des FRACTIONS de la boîte de la mosaïque.
 */
class MosaicLayoutTest {

    private data class Frame(val index: Int, val x: Float, val y: Float, val width: Float, val height: Float, val overflow: Int)

    private fun MosaicTile.rounded() = Frame(
        index = index,
        x = round3(x),
        y = round3(y),
        width = round3(width),
        height = round3(height),
        overflow = overflow,
    )

    private fun round3(value: Float): Float = Math.round(value * 1000f) / 1000f

    private fun tiled(count: Int, mode: MosaicLayoutMode): MosaicArrangement.Tiled =
        MosaicLayout.arrange(count, mode) as MosaicArrangement.Tiled

    @Test
    fun `hero a trois visuels pose une grande tuile de 0,62 et deux satellites en colonne`() {
        val arrangement = tiled(3, MosaicLayoutMode.HERO)

        assertThat(arrangement.tiles.map { it.rounded() }).containsExactly(
            Frame(0, 0f, 0f, 0.62f, 1f, 0),
            Frame(1, 0.634f, 0f, 0.366f, 0.493f, 0),
            Frame(2, 0.634f, 0.507f, 0.366f, 0.493f, 0),
        ).inOrder()
        assertThat(arrangement.aspectRatio).isEqualTo(0.82f)
        assertThat(arrangement.span).isEqualTo(1f)
    }

    @Test
    fun `wave a quatre visuels garde la largeur et fait onduler les hauteurs`() {
        assertThat(tiled(4, MosaicLayoutMode.WAVE).tiles.map { it.rounded() }).containsExactly(
            Frame(0, 0f, 0f, 0.24f, 1f, 0),
            Frame(1, 0.254f, 0.13f, 0.24f, 0.74f, 0),
            Frame(2, 0.507f, 0f, 0.24f, 1f, 0),
            Frame(3, 0.761f, 0.13f, 0.24f, 0.74f, 0),
        ).inOrder()
    }

    @Test
    fun `sine a trois visuels saute d'un bord a l'autre`() {
        val tiles = tiled(3, MosaicLayoutMode.SINE).tiles.map { it.rounded() }

        assertThat(tiles.map { it.y }).containsExactly(0f, 0.38f, 0f).inOrder()
        assertThat(tiles.map { it.height }.distinct()).containsExactly(0.62f)
    }

    @Test
    fun `reel deborde a droite au pas de 0,628 et sa boite defile`() {
        val arrangement = tiled(3, MosaicLayoutMode.REEL)

        assertThat(arrangement.tiles.map { listOf(round3(it.x), round3(it.width), round3(it.height)) })
            .containsExactly(listOf(0f, 0.6f, 1f), listOf(0.628f, 0.6f, 1f), listOf(1.256f, 0.6f, 1f))
            .inOrder()
        assertThat(round3(arrangement.span)).isEqualTo(1.856f)
    }

    @Test
    fun `chaque mosaique declare le rapport hauteur sur largeur de sa boite`() {
        assertThat(tiled(2, MosaicLayoutMode.WAVE).aspectRatio).isEqualTo(0.78f)
        assertThat(tiled(2, MosaicLayoutMode.HERO).aspectRatio).isEqualTo(0.82f)
        assertThat(tiled(2, MosaicLayoutMode.REEL).aspectRatio).isEqualTo(1.05f)
        assertThat(tiled(2, MosaicLayoutMode.SINE).aspectRatio).isEqualTo(0.92f)
    }

    @Test
    fun `au-dela de quatre les mosaiques comptent le reste sur la derniere tuile seule`() {
        val tiles = tiled(6, MosaicLayoutMode.HERO).tiles

        assertThat(tiles).hasSize(4)
        assertThat(tiles.map { it.overflow }).containsExactly(0, 0, 0, 2).inOrder()
    }

    @Test
    fun `le carrousel pagine tous les visuels sans rien cacher`() {
        assertThat(MosaicLayout.arrange(6, MosaicLayoutMode.CAROUSEL)).isEqualTo(MosaicArrangement.Paged(pageCount = 6))
    }

    /** Une mosaïque d'un élément n'en est pas une, quel que soit le mode. */
    @Test
    fun `aucun visuel ne pose rien et un seul visuel reste seul`() {
        MosaicLayoutMode.entries.forEach { mode ->
            assertThat(MosaicLayout.arrange(0, mode)).isEqualTo(MosaicArrangement.Empty)
            assertThat(MosaicLayout.arrange(1, mode)).isEqualTo(MosaicArrangement.Single)
        }
    }

    @Test
    fun `seules la grande tuile d'un hero, les tuiles d'un reel et une page portent une legende`() {
        assertThat(MosaicLayout.tileCarriesCaption(MosaicLayoutMode.HERO, index = 0, count = 3)).isTrue()
        assertThat(MosaicLayout.tileCarriesCaption(MosaicLayoutMode.HERO, index = 1, count = 3)).isFalse()
        assertThat(MosaicLayout.tileCarriesCaption(MosaicLayoutMode.REEL, index = 2, count = 3)).isTrue()
        assertThat(MosaicLayout.tileCarriesCaption(MosaicLayoutMode.CAROUSEL, index = 2, count = 3)).isTrue()
        assertThat(MosaicLayout.tileCarriesCaption(MosaicLayoutMode.WAVE, index = 0, count = 3)).isFalse()
        assertThat(MosaicLayout.tileCarriesCaption(MosaicLayoutMode.SINE, index = 0, count = 3)).isFalse()
        assertThat(MosaicLayout.tileCarriesCaption(MosaicLayoutMode.WAVE, index = 0, count = 1)).isTrue()
    }

    @Test
    fun `le budget de mots se reduit a la part de la rangee, jamais sous deux mots`() {
        assertThat(MosaicLayout.captionWordLimit(MosaicLayoutMode.HERO, count = 3, tileWidth = 0.62f)).isEqualTo(5)
        assertThat(MosaicLayout.captionWordLimit(MosaicLayoutMode.REEL, count = 3, tileWidth = 0.6f)).isEqualTo(12)
        assertThat(MosaicLayout.captionWordLimit(MosaicLayoutMode.WAVE, count = 4, tileWidth = 0.2395f)).isEqualTo(2)
        assertThat(MosaicLayout.captionWordLimit(MosaicLayoutMode.HERO, count = 1, tileWidth = 1f)).isEqualTo(20)
    }

    @Test
    fun `une legende se tronque en mots, jamais en lignes`() {
        assertThat(MosaicLayout.truncateWords("un deux trois quatre", limit = 2)).isEqualTo("un deux…")
        assertThat(MosaicLayout.truncateWords("  un   deux ", limit = 2)).isEqualTo("  un   deux ")
    }
}
